src/libcrun/linux.c

```diff
@@ -1947,6 +1947,160 @@ do_mounts (libcrun_container_t *container, int rootfsfd, const char *rootfs, con
+    }
+
+  if (mount_source && (flags & MS_BIND))
+    {
+      is_dir = crun_dir_p (mount_source, false, err);
+      if (UNLIKELY (is_dir < 0))
+        return is_dir;
+
+      data = append_mode_if_missing (data, "mode=1755");
+    }
+
+  /* Make sure any other directory/file is created and take a O_PATH reference to it.  */
+  ret = crun_safe_create_and_open_ref_at (is_dir, rootfsfd, rootfs, rootfs_len, target,
+                                          is_dir ? 01755 : 0755, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  targetfd = ret;
+
+  if (extra_flags & OPTION_TMPCOPYUP)
+    {
+      if (strcmp (type, "tmpfs") != 0)
+        return crun_make_error (err, 0, "tmpcopyup can be used only with tmpfs");
+
+      /* targetfd is opened with O_PATH, reopen the fd so it can read.  */
+      copy_from_fd = openat (targetfd, ".", O_RDONLY | O_DIRECTORY);
+      if (UNLIKELY (copy_from_fd < 0))
+        {
+          if (errno != ENOTDIR)
+            return crun_make_error (err, errno, "cannot reopen `%s`", target);
+
+          crun_error_release (err);
+        }
+    }
+
+  source = mount_source ? mount_source : type;
+
+  /* Check if there is already a mount for the requested file system.  */
+  if (mount_fd_index != -1)
+    {
+      if (mount_fds && mount_fds->fds[mount_fd_index] >= 0)
+        {
+          cleanup_close int mfd = get_and_reset (&(mount_fds->fds[mount_fd_index]));
+
+          ret = fs_move_mount_to (mfd, targetfd, NULL);
+          if (LIKELY (ret == 0))
+            {
+              /* Force no MS_BIND flag to not attempt again the bind mount.  */
+              ret = do_mount (container, NULL, mfd, target, NULL, flags & ~MS_BIND, data, LABEL_NONE, err);
+              if (UNLIKELY (ret < 0))
+                return ret;
+              mounted = true;
+            }
+        }
+    }
+
+  if (! mounted)
+    {
+      int label_how = LABEL_MOUNT;
+      ret = do_mount (container, source, targetfd, target, type, flags, data, label_how, err);
+      if (UNLIKELY (ret < 0))
+        return ret;
+    }
+
+  if (copy_from_fd >= 0)
+    {
+      int destfd, tmpfd;
+
+      destfd = safe_openat (rootfsfd, rootfs, rootfs_len, target, O_DIRECTORY, 0, err);
+      if (UNLIKELY (destfd < 0))
+        return crun_make_error (err, errno, "open target to write for tmpcopyup");
+
+      /* take ownership for the fd.  */
+      tmpfd = get_and_reset (&copy_from_fd);
+
+      ret = copy_recursive_fd_to_fd (tmpfd, destfd, target, target, err);
+      if (UNLIKELY (ret < 0))
+        return ret;
+    }
+
+  if (rec_clear || rec_set)
+    {
+      cleanup_close int dfd = -1;
+
+      dfd = safe_openat (rootfsfd, rootfs, rootfs_len, target, O_DIRECTORY, 0, err);
+      if (UNLIKELY (dfd < 0))
+        return crun_make_error (err, errno, "open mount target `/%s`", target);
+
+      ret = do_mount_setattr (target, dfd, rec_clear, rec_set, err);
+      if (UNLIKELY (ret < 0))
+        return ret;
+    }
+  return 0;
+}
+
```

Comment:
this looks like a lot of duplication we should avoid.

Could we instead have a simpler function that does just what we need (e.g. `libcrun_do_bind_mount` with source and dir paths)?

Reply:
@giuseppe Could you PTAL again.